### PR TITLE
PIM-9726: mitigate deadlock on categories (backport PIM-9548)

### PR DESCRIPTION
**Pull request by willy-ahva**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | OK
| Changelog updated                 | OK
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

**Review by BitOne — CHANGES_REQUESTED** [review]
Why do we apply a different solution than usual with deadlock?

File: upgrades/schema/Version_4_0_20210331092434_add_lock_table.php [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
+
+final class Version_4_0_20210331092434_add_lock_table extends AbstractMigration
+{
+    public function up(Schema $schema) : void
+    {
+        $sql = <<<SQL
+CREATE TABLE lock_keys (
```
Comment (ahocquard) [inline upgrades/schema/Version_4_0_20210331092434_add_lock_table.php:13]: To create on the fly:
```suggestion
CREATE TABLE IF NOT EXISTS lock_keys (
```

File: src/Akeneo/Pim/Enrichment/Component/Lock/Query/EnsureLockTableExistsInterface.php [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+namespace Akeneo\Pim\Enrichment\Component\Lock\Query;
+
+/**
+ * TODO pull up remove this interface
+ *
+ * @author    Willy Mesnage <willy.mesnage@akeneo.com>
+ * @copyright 2021 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+interface EnsureLockTableExistsInterface
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Component/Lock/Query/EnsureLockTableExistsInterface.php:14]: useless interface but not a big deal as it should not be pulled up

**Review by ahocquard — APPROVED** [review]
I would put directly the creation of the table inside the CategorySaver instead of wrapping it in a class + interface.

But as it will not be pull up it's ok.

**Review by BitOne — APPROVED** [review]

**Pull request merged**